### PR TITLE
Feature/vmp 227 add snapshot based otp mode

### PR DESCRIPTION
**Pull request by vessla**
When reading snapshotTimeout parameter with invalid value (that cannot be properly parsed to create LocalDateTime instance) into the RoutingRequest object I have used the same approach to "reporting" an error as in the case when user wants to combine rentingAllowed with arriveBy parameter. I have decided to do this not to introduce more changes to this PR. However I have noticed, that this approach (similarly to throwing ParameterException) generates 500 Internal Server Error instead of actually returning some meaningful information to the client due to the fact that it prevents some other variables from being initialized. It also seems to me, that there is something strange going on in this line in the PlannerResource class:
if(!PlannerError.isPlanningError(e.getClass()))
                LOG.warn("Error while planning path: ", e);

### CODE REVIEW

File: src/main/java/org/opentripplanner/standalone/CommandLineParameters.java [hunk 1]
```diff
@@ -146,9 +146,15 @@
...
     boolean enableScriptingWebService = false;
 
-    @Parameter(names = {"--sharedVehiclesApi"}, description = "API for fetching info about renable vehicles")
+    @Parameter(names = {"--sharedVehiclesApi"}, description = "API for fetching info about rentable vehicles")
     public String sharedVehiclesApi = null;
 
+    @Parameter(names = {"--sharedVehiclesHistoryApi"}, description = "API for fetching historical data about rentable vehicles")
+    public String sharedVehiclesHistoryApi = null;
+
+    @Parameter(names = {"--sharedVehiclesHistoryApiPass"}, description = "Password for fetching historical data about rentable vehicles")
```
Comment (AdamWiktor) [inline src/main/java/org/opentripplanner/standalone/CommandLineParameters.java]: Password should be passed only as environment variable

File: src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicles_positions/SharedHistoricalVehiclesUpdater.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                    LOG.error("Received empty provider list - waiting to try again");
+                    Thread.sleep(TimeUnit.SECONDS.toMillis(retryWaitTimeSeconds));
+                }
+            }
+            for (SharedVehiclesSnapshotLabel snapshotLabel : monitoredSnapshots) {
+                LOG.info("Polling vehicles from API for snapshot " + snapshotLabel.getTimestamp());
+                vehiclePositionsGetter = new VehicleHistoricalPositionsGetter(snapshotLabel, vehicleProviders);
+                vehiclePositionsGetter.setReturnNullOnNoResponse(true);
+                List<VehicleDescription> vehicles = null;
+                while (Objects.isNull(vehicles)) {
```
Comment (AdamWiktor) [inline src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicles_positions/SharedHistoricalVehiclesUpdater.java]: Maybe it should be failing after X tries?

File: src/main/java/org/opentripplanner/hasura_client/HasuraGetter.java [hunk 3]
```diff
@@ -1,17 +1,23 @@
...
 
 import java.util.List;
+import java.util.Objects;
 
 import static java.util.Collections.emptyList;
 
 public abstract class HasuraGetter<GRAPH_OBJECT, HASURA_OBJECT extends HasuraObject> {
+
+    @Setter
+    private boolean returnNullOnNoResponse = false;
```
Comment (AdamWiktor) [inline src/main/java/org/opentripplanner/hasura_client/HasuraGetter.java]: This should be final and set in a constructor. There might be a zero-argument constructor setting it to false and second one which gets this field as an argument

File: src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicles_positions/SharedHistoricalVehiclesUpdater.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+        this.url = Optional.ofNullable(System.getenv("SHARED_VEHICLES_HISTORY_UPDATER_API"))
+                .orElseGet(() -> System.getProperty("sharedVehiclesHistoryApi"));
+        if (Objects.isNull(this.url)) {
+            throw new IllegalStateException("Please provide program parameter `--sharedVehiclesHistoryApi <URL>`");
+        }
+
+        this.password = Optional.ofNullable(System.getenv("SHARED_VEHICLES_HISTORY_UPDATER_API_PASS"))
+                .orElseGet(() -> System.getProperty("sharedVehiclesHistoryApiPass"));
+        if (Objects.isNull(this.password)) {
```
Comment (AdamWiktor) [inline src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicles_positions/SharedHistoricalVehiclesUpdater.java]: Check if password is empty, this way we can run it in kubernetes and disable when we want. (Disabling will be done by setting url and/or password to empty string).

File: src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicles_positions/SharedHistoricalVehiclesUpdater.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    private GraphUpdaterManager graphUpdaterManager;
+    private Graph graph;
+    private String url;
+    private String password;
+    private int retryWaitTimeSeconds;
+    private final Map<Integer, Provider> vehicleProviders = new HashMap<>();
+    private final List<SharedVehiclesSnapshotLabel> monitoredSnapshots = new ArrayList<>();
+
+    @Override
+    protected void runPolling() {
```
Comment (AdamWiktor) [inline src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicles_positions/SharedHistoricalVehiclesUpdater.java:39]: This pooling and all runnables will take a really long time, maybe we should consider making it a graph builder in future?

**Review by AdamWiktor — APPROVED** [review]
There are many changes but all of them seems fine.

**Comment by AdamWiktor** [comment]
Proper handling of error codes is a topic for another ticket - I am aware that there are only few which are handled properly.

File: src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicles_positions/SharedHistoricalVehiclesUpdater.java (hunk 2 above)
Comment (vessla) [reply src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicles_positions/SharedHistoricalVehiclesUpdater.java]: I have added possibility to configure maximum number of retries in the router-config.json file (if the value is set to -1, updater will attempt to download the requested data until it receives a valid response)

**Pull request merged**